core/src/main/java/org/apache/cxf/io/CacheAndWriteOutputStream.java

```diff
@@ -47,19 +47,32 @@ public void setCacheLimit(long l) {
     }
 
     public void closeFlowthroughStream() throws IOException {
-        flowThroughStream.flush();
-        flowThroughStream.close();
+        postClose();
     }
 
     protected void postClose() throws IOException {
-        flowThroughStream.flush();
-        flowThroughStream.close();
+        if (!isClosed) {
+            flowThroughStream.flush();
+            flowThroughStream.close();
+            isClosed = true;
+        }
     }
 
     public OutputStream getFlowThroughStream() {
         return flowThroughStream;
     }
 
+    @Override
+    protected void doClose() throws IOException {
+        super.doClose();
+        isClosed = true;
```

Comment:
Just for the record, probably not necessary (since `close()` would call `postClose()`) but won't hurt

Reply:
@Sevyls I had to remove `close()` and `doClose()`, the complex relationships between these methods led to the issues that `isClosed` flag being set **before** closing the streams. As the results, streams stayed open.